messages.go

```diff
@@ -0,0 +1,140 @@
+package dht
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-libp2p-core/routing"
+	record "github.com/libp2p/go-libp2p-record"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	recpb "github.com/libp2p/go-libp2p-record/pb"
+
+	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
+)
+
+type ProtocolMessenger struct {
+	m         *messageManager
+	validator record.Validator
+}
+
+func NewProtocolMessenger(host host.Host, protocols []protocol.ID, validator record.Validator) *ProtocolMessenger {
+	return &ProtocolMessenger{
+		m: &messageManager{
+			host:      host,
+			strmap:    make(map[peer.ID]*messageSender),
+			protocols: protocols,
+		},
+		validator: validator,
+	}
+}
+
+// putValueToPeer stores the given key/value pair at the peer 'p'
+func (pm *ProtocolMessenger) PutValue(ctx context.Context, p peer.ID, rec *recpb.Record) error {
+	pmes := pb.NewMessage(pb.Message_PUT_VALUE, rec.Key, 0)
+	pmes.Record = rec
+	rpmes, err := pm.m.sendRequest(ctx, p, pmes)
+	if err != nil {
+		logger.Debugw("failed to put value to peer", "to", p, "key", loggableKeyBytes(rec.Key), "error", err)
+		return err
+	}
+
+	if !bytes.Equal(rpmes.GetRecord().Value, pmes.GetRecord().Value) {
```

Comment:
@aschmahmann Out of my own curiosity, why does this exist ?

Reply:
The reason the record values might be different is that A sends "here is the value I want B to store" and B's response is "here is the value I currently have". If those two aren't the same (e.g. A sent an out of date IPNS record) then an error can be propagated up the stack.

Aside from emitting log entries these errors aren't currently processed anywhere else.